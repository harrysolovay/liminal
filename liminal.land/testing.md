# Liminal Testing <Badge type="warning" text="beta" />
