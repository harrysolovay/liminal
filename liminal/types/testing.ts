import * as T from "./index.ts"

const g = T.union(
  T.object({
    a: T.number`Testing`,
    b: T.string,
    c: T.const(T.string, "hello"),
    d: {
      another: T.string,
      nested: T.object({
        e: T.number,
      }),
    },
    x: ["A", "B"],
  })`
  Testing
`,
  T.string,
)

console.log(JSON.stringify(g.toJSON(), null, 2))
