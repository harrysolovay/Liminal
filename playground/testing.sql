-- migrate:up
CREATE EXTENSION if NOT EXISTS "uuid-ossp";

CREATE EXTENSION if NOT EXISTS "hstore";

CREATE EXTENSION if NOT EXISTS "vector";

CREATE EXTENSION if NOT EXISTS "lo";

CREATE TABLE IF NOT EXISTS todo (
  id serial PRIMARY KEY,
  task TEXT,
  done BOOLEAN DEFAULT FALSE
);

INSERT INTO
  todo (task, done)
VALUES
  ('Install PGlite from NPM', TRUE);

INSERT INTO
  todo (task, done)
VALUES
  ('Load PGlite', TRUE);

INSERT INTO
  todo (task, done)
VALUES
  ('Create a table', TRUE);

INSERT INTO
  todo (task, done)
VALUES
  ('Insert some data', TRUE);

INSERT INTO
  todo (task)
VALUES
  ('Update a task');
