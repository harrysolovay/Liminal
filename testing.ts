import { ensureDir } from "@std/fs"
import * as path from "@std/path"
import { tap } from "./util/tap.ts"

export const dbg = tap(async (value) => {
  console.debug(value)
  await ensureDir(TMP_DIR)
  const dest = path.join(TMP_DIR, `${Date.now()}.json`)
  await Deno.writeTextFile(dest, JSON.stringify(value, null, 2))
  console.log(`Written to ${dest}.`)
})

const TMP_DIR = ".tmp"
