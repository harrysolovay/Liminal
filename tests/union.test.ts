import { assertSnapshot } from "@std/testing/snapshot"
import * as T from "../mod.ts"

Deno.test("No Variant Values", async (t) => {
  const Inner = T.union(
    "Inner description",
    [
      T.literal(undefined, "A"),
      T.literal(undefined, "B"),
    ],
  )
  const Root = T.object("Root description", { inner: Inner })
  await assertSnapshot(t, T.schema({ Root, Inner }, "Root"))
})

Deno.test("Variant Values", async (t) => {
  const Inner = T.union(
    "Inner description",
    [
      T.object(
        "A description",
        {
          type: T.literal(undefined, "A"),
          value: T.number(),
        },
      ),
      T.object(
        "B description",
        {
          type: T.literal(undefined, "B"),
          value: T.string(),
        },
      ),
      T.object(
        "C",
        { type: T.literal(undefined, "C") },
      ),
    ],
  )
  const Root = T.object("Root Description", { inner: Inner })
  await assertSnapshot(t, T.schema({ Root, Inner }, "Root"))
})
