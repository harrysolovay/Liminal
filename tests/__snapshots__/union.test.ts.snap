export const snapshot = {};

snapshot[`No Variant Values 1`] = `
{
  "\$defs": {
    Inner: {
      description: "Inner description",
      oneOf: [
        {
          const: "A",
          description: undefined,
        },
        {
          const: "B",
          description: undefined,
        },
      ],
    },
    Root: {
      additionalProperties: false,
      description: "Root description",
      properties: {
        inner: {
          "\$ref": "#/\$defs/Inner",
        },
      },
      required: [
        "inner",
      ],
      type: "object",
    },
  },
  additionalProperties: false,
  description: "Root description",
  properties: {
    inner: {
      "\$ref": "#/\$defs/Inner",
    },
  },
  required: [
    "inner",
  ],
  type: "object",
}
`;

snapshot[`Variant Values 1`] = `
{
  "\$defs": {
    Inner: {
      description: "Inner description",
      oneOf: [
        {
          additionalProperties: false,
          description: "A description",
          properties: {
            type: {
              const: "A",
              description: undefined,
            },
            value: {
              description: undefined,
              type: "number",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          description: "B description",
          properties: {
            type: {
              const: "B",
              description: undefined,
            },
            value: {
              description: undefined,
              type: "string",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          description: "C",
          properties: {
            type: {
              const: "C",
              description: undefined,
            },
          },
          required: [
            "type",
          ],
          type: "object",
        },
      ],
    },
    Root: {
      additionalProperties: false,
      description: "Root Description",
      properties: {
        inner: {
          "\$ref": "#/\$defs/Inner",
        },
      },
      required: [
        "inner",
      ],
      type: "object",
    },
  },
  additionalProperties: false,
  description: "Root Description",
  properties: {
    inner: {
      "\$ref": "#/\$defs/Inner",
    },
  },
  required: [
    "inner",
  ],
  type: "object",
}
`;
